### PR TITLE
Refs #28595 -- Added a hook to add execute wrappers for database queries.

### PR DESCRIPTION
**Pull request by shaib**
This is a (probably better) alternative to #9078, using an API inspired by middleware as suggested by @akaariai

### CODE REVIEW

**Comment by shaib** [comment]
buildbot, test on Oracle.

File: docs/topics/db/instrumentation.txt [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+For a more complete example, a logger of queries could look like this::
+
+    import time
+
+    class QueryLogger:
+
+        def __init__(self):
+            self.queries = []
+
+        def log(execute, sql, params, many, context):
```
Comment (adamchainz) [inline docs/topics/db/instrumentation.txt]: why not use `__call__` and then below the example can be just `with connection.execute_wrapper(QueryLogger())` ?

File: django/db/backends/utils.py [hunk 2]
```diff
@@ -65,14 +66,27 @@ def callproc(self, procname, params=None, kparams=None):
...
     def execute(self, sql, params=None):
+        return self._execute_with_wrappers(sql, params, False)
+
+    def executemany(self, sql, param_list):
+        return self._execute_with_wrappers(sql, param_list, True)
+
+    def _execute_with_wrappers(self, sql, params, many):
+        executor = self._execute if not many else self._executemany
+        context = dict(connection=self.db, cursor=self)
+        for wrapper in reversed(self.db.get_execute_wrappers(many)):
```
Comment (adamchainz) [inline django/db/backends/utils.py]: the `reversed` call isn't free, it's slightly more optimal to put the wrappers in the list in the way you want to iterate them

File: django/db/backends/utils.py [hunk 3]
```diff
@@ -65,14 +66,27 @@ def callproc(self, procname, params=None, kparams=None):
...
 
     def execute(self, sql, params=None):
+        return self._execute_with_wrappers(sql, params, False)
+
+    def executemany(self, sql, param_list):
+        return self._execute_with_wrappers(sql, param_list, True)
+
+    def _execute_with_wrappers(self, sql, params, many):
+        executor = self._execute if not many else self._executemany
+        context = dict(connection=self.db, cursor=self)
```
Comment (adamchainz) [inline django/db/backends/utils.py]: `{'connection': self.db, 'cursor': self}` is faster and imo clearer. `dict(foo=bar)` actually goes through two dict creation steps: first it creates the keyword args dict, then it finds the global name 'dict' (which could refer to anything), then it calls that name which creates the final dict.

File: docs/topics/db/instrumentation.txt [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        template_name = ...
+        with connection.execute_wrapper(blocker):
+            return render(request, template_name, context)
+
+The parameters sent to the wrappers are:
+
+* ``execute`` -- a callable, which should be invoked with the rest of the
+  parameters in order to execute the query;
+
+* ``sql`` -- the SQL text to be sent to the database engine;
```
Comment (adamchainz) [inline docs/topics/db/instrumentation.txt]: might be nice to make the types clear for all the args here, e.g. "a `str`, the SQL statement(s) to be sent..."

File: docs/topics/db/instrumentation.txt [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+* ``params`` -- parameters for the SQL statement (or statements);
```
Comment (adamchainz) [inline docs/topics/db/instrumentation.txt]: You can move the description of the type of `params` here, so it's harder to miss: "a list of parameters for the SQL statement(s), or if `many` is true, a list of such lists, one for each execution."

File: docs/topics/db/instrumentation.txt [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+  invocation. For now, this includes the connection and cursor.
+
+So, a slightly more complex version of the blocker could include the connection
+name in the error message::
+
+    def blocker(execute, sql, params, many, context):
+        alias = context['connection'].alias
+        raise Exception("Access to database '{}' blocked here".format(alias))
+
+For a more complete example, a logger of queries could look like this::
```
Comment (adamchainz) [inline docs/topics/db/instrumentation.txt]: "a query logger could look like..." ? 🌴

File: docs/topics/db/instrumentation.txt [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+    import time
+
+    class QueryLogger:
+
+        def __init__(self):
+            self.queries = []
+
+        def log(execute, sql, params, many, context):
+            current_query = [sql, params, many]
```
Comment (adamchainz) [inline docs/topics/db/instrumentation.txt]: I think this would be easier to follow if `current_query` were a dict, then you would just add keys to it below, rather than all these `append` operations which are a bit hard to follow with the multiple execution paths and a full try/except/else/finally statement

File: tests/backends/base/test_base.py [hunk 8]
```diff
@@ -30,3 +32,113 @@ def test_initialization_class_attributes(self):
...
+        with connection.cursor() as cursor:
+            cursor.execute(sql, params)
+
+    def call_executemany(self, connection, params_list=((1,),) * 3):
+        sql = 'GIBBERISH with SELECT in it %s'
+        # We need a no-op query, but it has to be DML. We could use a model
+        # to make sure there's a table we can update, or we could just catch
+        # the error.
+        with self.assertRaises(Exception):
+            with connection.cursor() as cursor:
```
Comment (adamchainz) [inline tests/backends/base/test_base.py]: you can collapse, `with self.assertRaises(Exception), connection.cursor() as cursor:`, and in a few places below.

File: docs/topics/db/instrumentation.txt [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+    def my_view(request):
+        context = ...  # Code to generate context with all data
+        template_name = ...
+        with connection.execute_wrapper(blocker):
+            return render(request, template_name, context)
+
+The parameters sent to the wrappers are:
+
+* ``execute`` -- a callable, which should be invoked with the rest of the
+  parameters in order to execute the query;
```
Comment (adamchainz) [inline docs/topics/db/instrumentation.txt]: I think the semicolons are best replaced as full stops / periods.

File: docs/topics/db/instrumentation.txt [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+
+* ``sql`` -- the SQL text to be sent to the database engine;
+
+* ``params`` -- parameters for the SQL statement (or statements);
+
+* ``many`` -- a Boolean indicating whether the ultimately invoked call is
```
Comment (adamchainz) [inline docs/topics/db/instrumentation.txt]: "a `bool`" ?

File: docs/topics/db/instrumentation.txt [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+  ``execute()`` or ``executemany()`` (and whether ``params`` is expected to be
+  a list of values, or a list of lists of values);
+
+* ``context`` -- a dictionary with further data about the context of
+  invocation. For now, this includes the connection and cursor.
+
+So, a slightly more complex version of the blocker could include the connection
```
Comment (adamchainz) [inline docs/topics/db/instrumentation.txt]: "So" feels a bit unnecessary/colloquial here.

File: docs/topics/db/instrumentation.txt (hunk 5 above)
Comment (adamchainz) [reply docs/topics/db/instrumentation.txt]: In fact, considering the test failure, perhaps "a list/tuple of parameters" :)

File: django/db/backends/utils.py (hunk 3 above)
Comment (shaib) [reply django/db/backends/utils.py]: Yes, I like using `dict` when the keys are really symbols, but you're right.

File: docs/topics/db/instrumentation.txt (hunk 9 above)
Comment (shaib) [reply docs/topics/db/instrumentation.txt]: Accepted.

File: docs/topics/db/instrumentation.txt (hunk 4 above)
Comment (shaib) [reply docs/topics/db/instrumentation.txt]: Three had the info already, added for the other two.

File: docs/topics/db/instrumentation.txt (hunk 5 above)
Comment (shaib) [reply docs/topics/db/instrumentation.txt]: I phrased a little differently, but accepted the idea.

File: docs/topics/db/instrumentation.txt (hunk 10 above)
Comment (shaib) [reply docs/topics/db/instrumentation.txt]: Yep.

File: docs/topics/db/instrumentation.txt (hunk 11 above)
Comment (shaib) [reply docs/topics/db/instrumentation.txt]: Replaced with "Using them," referring to the parameters above

File: docs/topics/db/instrumentation.txt (hunk 6 above)
Comment (shaib) [reply docs/topics/db/instrumentation.txt]: Of course. Thanks.

File: docs/topics/db/instrumentation.txt (hunk 1 above)
Comment (shaib) [reply docs/topics/db/instrumentation.txt]: We need to keep a handle on the logger if we want access to the logged queries. We could have the context-manager return the wrapper from its `__enter__()`, and then we could have `with connection.execute_wrapper(QueryLogger()) as logger:`; but I'm not sure that the wrapper is the only useful thing to return from `__enter__()`, and since the wrapper is available to the programmer anyway, I'd like to avoid committing to returning it at this point.

Other than that, this is a very good idea, and lets us demonstrate what a callable that isn't a function looks like in this context.

File: docs/topics/db/instrumentation.txt (hunk 7 above)
Comment (shaib) [reply docs/topics/db/instrumentation.txt]: Good idea.

File: django/db/backends/utils.py (hunk 2 above)
Comment (shaib) [reply django/db/backends/utils.py]: The differences are infinitesimal (on my system, each `reversed()` call takes ~0.2 microseconds). I think it's more readable and debuggable this way -- when the wrappers are held in the order installed.

Note also that `append()` is significantly faster than `insert(0)` -- we expect quite short lists here, so it doesn't really matter, but in general, unless you read the list many times, append+reversed is the way to go:
```python
In [3]: %timeit x=[]; [x.append(i) for i in range(30)]; reversed(x)
4.41 µs ± 203 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [4]: %timeit x=[]; [x.insert(0,i) for i in range(30)]; 
6.28 µs ± 31.7 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [5]: %timeit x=[]; [x.append(i) for i in range(30)]; reversed(x); reversed(x); reversed(x)
4.77 µs ± 52.3 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```
It seems that for a list of length 30, you'd need to read it more than 20 times for `insert(0)` to pay off.

File: tests/backends/base/test_base.py (hunk 8 above)
Comment (shaib) [reply tests/backends/base/test_base.py]: Yep.

**Comment by shaib** [comment]
buildbot, test on Oracle.

File: docs/topics/db/instrumentation.txt (hunk 1 above)
Comment (adamchainz) [reply docs/topics/db/instrumentation.txt]: Good point, it doesn't make sense to not have a handle in this case. And yes no need to return it from the context manager.

File: django/db/backends/utils.py (hunk 2 above)
Comment (adamchainz) [reply django/db/backends/utils.py]: fair enough, I guess I'm just a bit too obsessed with optimization :)

Comment (adamchainz) [reply django/db/backends/utils.py]: btw `reversed(x)` doesn't actually iterate the whole list in reverse in python 3, you just get a `list_reverseiterator`...

```
In [1]: reversed([1])
Out[1]: <list_reverseiterator at 0x105098b70>
```

:)

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by shaib** [comment]
buildbot, test on oracle.

**Comment by shaib** [comment]
buildbot, test on oracle.

**Comment by holvianssi** [comment]
The feature looks really good to me.

I would have some use for globally installed hooks (measure all queries done by celery etc), but this is already an excellent addition, and a global version can be added easily enough in later versions.

All in all, +1 from me for the feature.

**Comment by shaib** [comment]
@holvianssi Thanks!
BTW, I have already enabled a similar feature (based on the design of #9078) in a project on Django 1.8, using custom database backends. I plan to release something similar as an independent package, which should be able to support any Django>=1.8. I haven't thought about global hooks at all, but I'm guessing we'll be able to think it through in that context.

**Comment by adamchainz** [comment]
One can implement global hooks with this by using a subclassed database backend that adds a hook after `__init__`, at least 😁

File: docs/topics/db/instrumentation.txt [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+
+To use this effectively, you would create a logger object and install it as a
+wrapper::
+
+    from django.db import connection
+
+    ql = QueryLogger()
+    with connection.execute_wrapper(ql):
+        # Do some interesting querying
+    # Access ql.queries here
```
Comment (adamchainz) [inline docs/topics/db/instrumentation.txt]: you could make this valid python by calling fake functions instead like `do_some_queries()` and `print(ql.queries)`

**Comment by shaib** [comment]
> After a bit of thought I'm slightly suspicious of whether the `for_many` flag
> is really needed. 

After a bit of thought, I agree.

The original idea was that using this flag could guarantee that a wrapper would 
always be called with parameters of the same type, and therefore could allow 
it to be written more cleanly.

But on second thought, a wrapper could be defined once and installed many 
times, and if it is written to rely on being installed with a specific 
`for_many` value, it becomes easy to misuse. This point marks the flag as, 
essentially, a footgun.

I'll now remove it.

**Comment by shaib** [comment]
Addressed the last comments, rebased and squashed.

**Comment by shaib** [comment]
buildbot, test on oracle.

**Comment by adamchainz** [comment]
> a wrapper could be defined once and installed many
> times, and if it is written to rely on being installed with a specific
> `for_many` value, it becomes easy to misuse. This point marks the flag as,
> essentially, a footgun.

Ah yes, hadn't even thought of that!

File: django/db/backends/base/base.py [hunk 13]
```diff
@@ -98,6 +98,15 @@ def __init__(self, settings_dict, alias=DEFAULT_DB_ALIAS,
         self.introspection = self.introspection_class(self)
         self.ops = self.ops_class(self)
         self.validation = self.validation_class(self)
+        # A stack of wrappers to be invoked around execute()/executemany()
+        # calls.
+        #
+        # Each entry is a (wrapper, for_many) tuple, where wrapper is a
```
Comment (alvra) [inline django/db/backends/base/base.py]: Is this comment still accurate regarding `for_many`?

Comment (adamchainz) [reply django/db/backends/base/base.py]: 🦅 eagle eye award

File: docs/topics/db/instrumentation.txt [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+For a more complete example, a query logger could look like this::
+
+    import time
+
+    class QueryLogger:
+
+        def __init__(self):
+            self.queries = []
+
+        def __call__(execute, sql, params, many, context):
```
Comment (Stranger6667) [inline docs/topics/db/instrumentation.txt]: `self` seems to be missed, since it is not a `staticmethod`

File: django/db/backends/base/base.py (hunk 13 above)
Comment (shaib) [reply django/db/backends/base/base.py]: Thanks, fixed.

File: docs/topics/db/instrumentation.txt (hunk 14 above)
Comment (shaib) [reply docs/topics/db/instrumentation.txt]: Thanks, corrected.

**Comment by shaib** [comment]
buildbot, test on oracle.

File: django/db/backends/utils.py [hunk 15]
```diff
@@ -65,14 +66,26 @@ def callproc(self, procname, params=None, kparams=None):
                 return self.cursor.callproc(procname, params, kparams)
 
     def execute(self, sql, params=None):
+        return self._execute_with_wrappers(sql, params, False, self._execute)
```
Comment (timgraham) [inline django/db/backends/utils.py]: It would be more readable to pass boolean arguments as keywords, e.g. `many=False`.

File: docs/topics/db/instrumentation.txt [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+
+.. method:: execute_wrapper(wrapper)
+
+Returns a context manager which, when entered, installs a wrapper around
+database query executions, and when exited, removes the wrapper. The wrapper is
+installed on the thread-local connection object.
+
+``wrapper`` is a callable taking five arguments.  It is called for every query
+execution in the scope of the context manager, with arguments ``execute``,
+``sql``, ``params``, ``many`` and ``context`` as described above. It is
```
Comment (timgraham) [inline docs/topics/db/instrumentation.txt]: , and

File: docs/topics/db/instrumentation.txt [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+
+    from django.db import connection
+
+    ql = QueryLogger()
+    with connection.execute_wrapper(ql):
+        do_interesting_querying()
+    # Now we can print the log:
+    report_queries(ql.queries)
+
+.. currentmodule:: django.db.backends.base.DatabaseWrapper
```
Comment (timgraham) [inline docs/topics/db/instrumentation.txt]: I wonder if we should consider starting a docs/ref document for public `DatabaseWrapper` methods. I'm just thinking about that since I noticed you wrote "as described above" below, which would have to be rewritten later if we did that. The other documented method that immediately comes to mind is `cursor()` (in `topics/db/sql.txt`).

File: tests/backends/base/test_base.py [hunk 18]
```diff
@@ -30,3 +34,96 @@ def test_initialization_class_attributes(self):
...
+
+    @staticmethod
+    def call_execute(connection, params=None):
+        ret_val = '1' if params is None else '%s'
+        sql = 'SELECT ' + ret_val + connection.features.bare_select_suffix
+        with connection.cursor() as cursor:
+            cursor.execute(sql, params)
+
+    def call_executemany(self, connection, params=None):
+        # Executemany must use an updating query, we make sure ours does nothing
```
Comment (timgraham) [inline tests/backends/base/test_base.py]:
```
# executemany() must use an update query. Make sure it does nothing
# by putting a false condition in the WHERE clause.
```

File: django/db/backends/base/base.py [hunk 19]
```diff
@@ -525,6 +532,26 @@ def validate_thread_sharing(self):
                 % (self.alias, self._thread_ident, _thread.get_ident())
             )
 
+    # ##### Execute wrappers #####
+
+    def _push_execute_wrapper(self, wrapper):
```
[the remark this answers is not in the document]
Comment (shaib) [reply django/db/backends/base/base.py]: Right. I'm removing them, and with that removal, the whole "Execute Wrappers" block becomes just one method, so I'm moving it down into the "miscellaneous" block.

File: django/db/backends/utils.py (hunk 15 above)
Comment (shaib) [reply django/db/backends/utils.py]: Added.

File: docs/topics/db/instrumentation.txt (hunk 17 above)
Comment (shaib) [reply docs/topics/db/instrumentation.txt]: My first and foremost doubt about this documentation was the location. I started a  new file because I couldn't find an existing place which felt right, and I agree that if there were an existing file for working with a connection object, that would probably be it. 

But I also suspect that creating that file is a little out of scope for this PR. Can you suggest a phrasing that would work for both the current arrangement and for a future arrangement where the `DatabaseWrapper` reference is separate? If not, I think we can just change the wording then. Perhaps leave a comment about it now.

One idea I considered was just to move the description of the arguments into the "reference" part, where it really belongs; but I can't think of a good way to do that without making the example (`QueryLogger`) a lot less readable.

File: docs/topics/db/instrumentation.txt (hunk 16 above)
Comment (shaib) [reply docs/topics/db/instrumentation.txt]: (damn foreign languages. In Hebrew, such commas as very bad style)

File: tests/backends/base/test_base.py (hunk 18 above)
Comment (shaib) [reply tests/backends/base/test_base.py]: Accepted

File: docs/topics/db/instrumentation.txt (hunk 16 above)
Comment (adamchainz) [reply docs/topics/db/instrumentation.txt]: They can be very important: https://www.nytimes.com/2017/03/16/us/oxford-comma-lawsuit.html 😂
